# Исключения
- [runCatching](#runcatching)
## runCatching
runCatching — это функция в Kotlin, которая позволяет безопасно обрабатывать исключения. Она предоставляет удобный 
способ выполнения кода, который может выбросить исключение, и позволяет работать с результатом либо успеха, либо
неудачи (исключения).

runCatching оборачивает выполнение блока кода в попытку поймать исключения (try-catch) и возвращает результат в виде
объекта Result<T>, который может содержать либо успешное значение (Success), либо информацию о возникшем исключении 
(Failure).
### Синтаксис
```kotlin
val result = runCatching {
    // Некоторый код, который может выбросить исключение
    performOperation()
}
```
Здесь, если performOperation() выполнится успешно, результат будет обёрнут в объект Success. Если возникнет исключение,
оно будет захвачено и сохранено в объекте Failure.
### Работа с результатом
- Результат выполнения runCatching можно обработать с помощью функций onSuccess и onFailure, которые позволяют задать
поведение для каждого исхода:
```kotlin
val result = runCatching {
    performOperation()
}

result
    .onSuccess { value ->
        println("Операция прошла успешно: $value")
    }
    .onFailure { exception ->
        println("Произошло исключение: ${exception.message}")
    }
```
- getOrNull: Возвращает результат, если операция была успешной, или null, если произошла ошибка:
```kotlin
val result = runCatching { divide(10, 0) }.getOrNull()
println(result)  // null
```
- getOrElse: Возвращает результат, если операция была успешной, или заданное значение, если произошла ошибка:
```kotlin
val result = runCatching { divide(10, 0) }.getOrElse { -1 }
println(result)  // -1
```
- recover: Позволяет вернуть альтернативное значение в случае ошибки:
```kotlin
val result = runCatching { divide(10, 0) }
    .recover { -1 }
println(result.getOrNull())  // -1
```
### Преимущества использования
1) Удобное управление ошибками: Вместо явного использования try-catch блоков можно более декларативно обрабатывать
результаты выполнения.
2) Чистый код: Функция позволяет сделать код более читабельным, особенно при цепочке вызовов, когда важно обработать
как успешные, так и ошибочные случаи.
3) Безопасность: Обработка ошибок с использованием runCatching снижает вероятность пропуска исключений и помогает 
писать более надёжный код.
